server/src/main/java/org/opensearch/common/collect/LinkedDeque.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.collect;
+
+import java.util.AbstractCollection;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Linked list implementation of the {@link Deque} interface where the link
+ * pointers are tightly integrated with the element. Linked deques have no
+ * capacity restrictions; they grow as necessary to support usage. They are not
+ * thread-safe; in the absence of external synchronization, they do not support
+ * concurrent access by multiple threads. Null elements are prohibited.
+ * <p>
+ * Most LinkedDeque operations run in constant time by assuming that
+ * the {@link Linked} parameter is associated with the deque instance. Any usage
+ * that violates this assumption will result in non-deterministic behavior.
+ * <p>
+ * The iterators returned by this class are <em>not</em> <i>fail-fast</i>: If
+ * the deque is modified at any time after the iterator is created, the iterator
+ * will be in an unknown state. Thus, in the face of concurrent modification,
+ * the iterator risks arbitrary, non-deterministic behavior at an undetermined
+ * time in the future.
+ *
+ * @param <E> the type of elements held in this collection
+ *
+ * @opensearch.internal
+ */
+public final class LinkedDeque<E extends Linked<E>> extends AbstractCollection<E> implements Deque<E> {
```

Comment:
Do you know where this implementation comes from? It seems to be [widely copied into projects on GitHub](https://github.com/search?q=%22This%20class%20provides%20a%20doubly-linked%20list%20that%20is%20optimized%20for%20the%20virtual%22&type=code).

Reply:
Really what I'm getting at is whether we need to implement this data structure or if we can use something built into the JDK instead.

Reply:
> Do you know where this implementation comes from? It seems to be [widely copied into projects on GitHub](https://github.com/search?q=%22This%20class%20provides%20a%20doubly-linked%20list%20that%20is%20optimized%20for%20the%20virtual%22&type=code).

No, I don't know where this file originally came from. But I see almost all cache implementations (Guava, Caffeine, ...) have this and no I'm not aware of any Java built in alternative

Reply:
@andrross this copied from [1], which is the predecessor of [2], we should not be doing  that (IMHO)

[1] https://github.com/ben-manes/concurrentlinkedhashmap/
[2] https://github.com/ben-manes/caffeine

Reply:
@reta There's now more details about why we're doing this in #4964. I would love to find a library implementation we can use for this and the caching logic in general, but at the moment we don't have one and I'd like to keep making incremental progress here. #6225 has been opened to track that effort.

Also, for additional context, this and other code in this PR has been adapted from UltraWarm (see [the commit message](https://github.com/opensearch-project/OpenSearch/pull/5641/commits/8fb7a39aac06d5067a3d8cbd450bd89370066c0e)) which is another reason why this implementation is our first approach.

What do you think?

Reply:
@andrross sure, have a suggestion here (https://github.com/opensearch-project/OpenSearch/pull/5641#issuecomment-1419665814) on isolating this code (in separate library)